File: older-code/inaimathi/sudoku/sudoku-elegant.py
#!/usr/bin/python
"""
Transliteration of Maggesi's tiny haskell sudoku solver: 
  http://web.math.unifi.it/users/maggesi/haskell_sudoku_solver.html

Not the ideal way to solve sudoku, but its small and elegant.

A board is represented by a function that maps coordinates to
possible values.  

The mark function returns a new board with a particular coordinate 
filled in and conflicts removed from related rows, columns and lines.

The solve function goes through and makes new possible boards 
by making guesses from the set of possibilities for each square.

All solutions are collected and printed.

Note: reductions are backwards because order wasn't important
and python comes with a left fold function but not a right fold.
"""

import sys

rs = range(1,10)
idx = [(i,j) for i in rs for j in rs]

def input(s) :
    def allVals(p) :
        return rs
    return reduce(mark, ((p,n) for p,n in zip(idx, (int(w) for l in s.split('\n') for w in l.split(' '))) if n > 0), allVals)

def disp(s) :
    return '\n'.join(' '.join(str(s((i,j))[0]) for j in rs) for i in rs)

def e(a, b) :
    return (a - 1) / 3 == (b - 1) / 3

def mark(s, pn) :
    p,n = pn
    i,j = p
    def filtered(q) :
        x,y = q
        if p == q :
            return [n]
        elif x == i or y == j or (e(x, i) and e(y, j)) :
            return [x for x in s(q) if x != n]
        else :
            return s(q)
    return filtered

def solve(s) :
    def search(l, p) :
        return [mark(s2, (p, n)) for s2 in l for n in s2(p)]
    return reduce(search, idx, s)

def main() :
    for fn in sys.argv[1:] :
        s = file(fn).read()[:-1]
        print '\n'.join(map(disp, solve([input(s)])))

if __name__ == '__main__' :
    main()

File: older-code/inaimathi/sudoku/sudoku-elegant.hs
-- Pilfered (but modified) from http://web.math.unifi.it/users/maggesi/haskell_sudoku_solver.html
import Data.List

type Board = (Int,Int) -> [Int]

main :: IO ()
main = putStr $ unlines $ map disp $ solve [input sampleFiendish]

----- Take a list of boards and return a list of solutions
solve :: [Board] -> [Board]
solve boards = foldr search boards idx

search :: (Int, Int) -> [Board] -> [Board]
search (x, y) boards = [mark ((x, y), val) b | b <- boards, val <- b (x, y)]

----- Take a (x, y)
mark :: ((Int,Int),Int) -> Board -> Board
mark (a@(x,y), val) board b@(x',y')
  | a == b = 
    [val]
  | x' == x || y' == y || (e x' x) && (e y' y) =
    delete val $ board b
  | otherwise = 
    board b
  where e a b = div (a-1) 3 == div (b-1) 3

----- Converting [[Int]] to ... something using mark
----- Looks like this is building up thunks? So that the result is going to be a function that takes a particular (x, y) tuple and
----- returns a list of possibilities for that board square, passing it through all previous deletions?
input :: [[Int]] -> Board
input s = foldr mark (const [1..9]) $                                  -- (const [1..9]) is equivalent to (\_ -> [1..9])
          [((x, y), val) | ((x, y), val) <- zip idx $ concat s, val>0] -- 0 is the empty value

----- Print a board
disp :: Board -> String
disp s  = unlines [unwords [show $ head $ s (i,j) | j <- [1..9]] | i <- [1..9]]

----- Each grid x,y pair
idx :: [(Int,Int)]
idx = [(x,y) | y <- [1..9], x <- [1..9]]

sample :: [[Int]]
sample = [[0,7,1,4,0,0,0,0,5],
          [0,0,0,0,5,0,0,8,0],
          [0,0,3,9,0,7,6,0,0],
          [0,0,0,0,0,1,0,0,0],
          [0,9,0,8,0,6,0,0,3],
          [0,0,0,0,0,0,8,2,0],
          [0,6,0,0,4,0,7,0,8],
          [3,0,0,0,0,0,0,9,0],
          [0,0,0,0,8,5,0,0,0]]

sampleHard :: [[Int]]
sampleHard = [[0,7,1,4,0,0,0,0,5],
              [0,0,0,0,5,0,0,8,0],
              [0,0,3,9,0,7,6,0,0],
              [0,0,0,0,0,1,0,0,0],
              [0,9,0,0,0,6,0,0,3],
              [0,0,0,0,0,0,8,2,0],
              [0,0,0,0,4,0,0,0,8],
              [3,0,0,0,0,0,0,9,0],
              [0,0,0,0,8,5,0,0,0]]

sampleFiendish :: [[Int]]
sampleFiendish = [[0,0,0,0,0,0,0,0,0],
                  [0,0,0,0,0,0,0,0,0],
                  [0,0,0,0,0,0,0,0,0],
                  [0,0,0,0,0,0,0,0,0],
                  [0,0,0,0,0,6,0,0,3],
                  [0,0,0,0,0,0,0,0,0],
                  [0,0,0,0,0,0,0,0,0],
                  [0,0,0,0,0,0,0,0,0],
                  [0,0,0,0,0,0,0,0,0]]